### PR TITLE
Convert match

### PR DESCRIPTION
**Pull request by shayne-fletcher-da**
This PR updates dlint to include the GHC version of "match rules". Tests affected by the newly enabled diagnostics are updated to expect them. Depends on PR https://github.com/digital-asset/hlint/pull/40.

### CODE REVIEW

File: compiler/damlc/tests/daml-test-files/List.daml [hunk 1]
```diff
@@ -1,6 +1,15 @@
 -- Copyright (c) 2019, Digital Asset (Switzerland) GmbH and/or its affiliates.
 -- All rights reserved.
 
+-- @INFO Use null
+-- @INFO Evaluate
+-- @INFO Use sum
```
Comment (hurryabit) [inline compiler/damlc/tests/daml-test-files/List.daml]: You might need to add explicit location annotations to make this work. IIRC, something about them feels like being off by one.

File: compiler/damlc/tests/daml-test-files/PartyCompare.daml [hunk 2]
```diff
@@ -8,14 +8,14 @@ test = scenario do
   alice <- getParty "Alice"
   bob <- getParty "Bob"
   assert $ alice < bob
-  assert $ not (alice < alice)
-  assert $ not (bob < alice)
+  assert $ alice >= alice
```
Comment (hurryabit) [inline compiler/damlc/tests/daml-test-files/PartyCompare.daml]: It looks like the intent of this test is to verify that `alice < alice` does not hold. Changing it would then be the wrong thing to do. I assume that's true for the whole file.

File: compiler/damlc/tests/daml-test-files/List.daml (hunk 1 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/tests/daml-test-files/List.daml]: I avoided the off by one problem by not adding ranges. Doing the minimum possible to get the tests through.

File: compiler/damlc/tests/daml-test-files/PartyCompare.daml (hunk 2 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/tests/daml-test-files/PartyCompare.daml]: Ah ok. In this instance I did change the test rather than just ignore the diagnostics. I'll fix it.

File: compiler/damlc/daml-ide-core/dlint.yaml [hunk 3]
```diff
@@ -276,13 +246,13 @@
     - warn: {lhs: f (fst p) (snd p), rhs: uncurry f p}
     - warn: {lhs: ($) . f, rhs: f, name: Redundant $}
     - warn: {lhs: (f $), rhs: f, name: Redundant $}
-    - warn: {lhs: (Data.Function.& f), rhs: f, name: Redundant Data.Function.&}
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/dlint.yaml]: Afaik we don’t have `&` in DAML. I guess this rule doesn’t do any harm in that case but it’s a bit confusing.

File: compiler/damlc/daml-ide-core/dlint.yaml [hunk 4]
```diff
@@ -276,13 +246,13 @@
...
+    - warn: {lhs: (& f), rhs: f, name: Redundant &}
     - hint: {lhs: \x -> y, rhs: const y, side: isAtom y && not (isWildcard y)}
         # If any isWildcard recursively then x may be used but not mentioned explicitly
     - warn: {lhs: flip f x y, rhs: f y x, side: isApp original}
-    - warn: {lhs: id x, rhs: x, side: not (isTypeApp x)}
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/dlint.yaml]: We do have type applications in DAML so why did you remove this check?

File: compiler/damlc/daml-ide-core/dlint.yaml [hunk 5]
```diff
@@ -330,36 +300,47 @@
...
+    - warn: {lhs: first f (second g x), rhs: bimap f g x}
+    - warn: {lhs: second g (first f x), rhs: bimap f g x}
+    - warn: {lhs: first f1 (first f2 x), rhs: first (f1 . f2) x}
+    - warn: {lhs: second g1 (second g2 x), rhs: second (g1 . g2) x}
+    - warn: {lhs: bimap f1 f2 (bimap g1 g2 x), rhs: bimap (f1 . f2) (g1 . g2) x}
+    - warn: {lhs: first f1 (bimap f2 g x), rhs: bimap (f1 . f2) g x}
+    - warn: {lhs: second g1 (bimap f g2 x), rhs: bimap f (g1 . g2) x}
+    - warn: {lhs: bimap f1 g (first f2 x), rhs: bimap (f1 . f2) g x}
+    - warn: {lhs: bimap f g1 (second g2 x), rhs: bimap f (g1 . g2) x}
+    - hint: {lhs: "\\(x,y) -> (f x, g y)", rhs: DA.Bifunctor.bimap f g, note: IncreasesLaziness}
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/dlint.yaml]: `IncreasesLaziness` doesn’t make sense for DAML.

File: compiler/damlc/daml-ide-core/dlint.yaml [hunk 6]
```diff
@@ -477,46 +458,48 @@
...
     # MAYBE
 
-    - warn: {lhs: maybe x id, rhs: Data.Maybe.fromMaybe x}
-    - warn: {lhs: maybe False (const True), rhs: Data.Maybe.isJust}
-    - warn: {lhs: maybe True (const False), rhs: Data.Maybe.isNothing}
-    - warn: {lhs: maybe False (== x), rhs: (== Just x)}
-    - warn: {lhs: maybe True (/= x), rhs: (/= Just x)}
-    - warn: {lhs: not (isNothing x), rhs: isJust x}
-    - warn: {lhs: not (isJust x), rhs: isNothing x}
+    - warn: {lhs: maybe x identity, rhs: DA.Optional.fromOptional x}
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/dlint.yaml]: All these `maybe`s should be `optional`.

File: compiler/damlc/daml-ide-core/dlint.yaml [hunk 7]
```diff
@@ -761,11 +703,11 @@
     rules:
     - warn: {lhs: maybe mempty, rhs: foldMap}
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/dlint.yaml]:
```suggestion
    - warn: {lhs: optional mempty, rhs: foldMap}
```

File: compiler/damlc/daml-ide-core/dlint.yaml [hunk 8]
```diff
@@ -761,11 +703,11 @@
...
     - warn: {lhs: either (const mempty), rhs: foldMap}
-    - warn: {lhs: Data.Maybe.fromMaybe mempty, rhs: Data.Foldable.fold}
-    - warn: {lhs: Data.Either.fromRight mempty, rhs: Data.Foldable.fold}
-    - warn: {lhs: if f x then Just x else Nothing, rhs: mfilter f (Just x)}
+    - warn: {lhs: DA.Optional.fromOptional mempty, rhs: DA.Foldable.fold}
+    - warn: {lhs: DA.Either.fromRight mempty, rhs: DA.Foldable.fold}
+    - warn: {lhs: if f x then Some x else None, rhs: mfilter f (Some x)}
     - hint: {lhs: maybe (pure ()), rhs: traverse_, note: IncreasesLaziness}
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/dlint.yaml]:
```suggestion
    - hint: {lhs: optional (pure ()), rhs: traverse_, note: IncreasesLaziness}
```

File: compiler/damlc/tests/daml-test-files/MaybeCompat.daml [hunk 9]
```diff
@@ -1,11 +1,13 @@
 -- Copyright (c) 2019, Digital Asset (Switzerland) GmbH and/or its affiliates.
 -- All rights reserved.
 
--- @WARN range=14:12-14:17; Maybe
--- @WARN range=17:12-17:17; maybe
--- @WARN range=17:29-17:36; Nothing
--- @WARN range=18:12-18:20; fromSome
--- @WARN range=18:22-18:26; Just
+-- @INFO Use fromOptional
+-- @WARN Maybe
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/MaybeCompat.daml]: Please don’t remove ranges that we already have. That makes it easy to introduce regressions in our location info.

File: stack-snapshot.yaml [hunk 10]
```diff
@@ -5,9 +5,9 @@ packages:
   - archive: http://digitalassetsdk.bintray.com/ghc-lib/ghc-lib-8.8.1.20191204.tar.gz
     sha256: "9648145ef7b3566fad7073ef63d0ef85056abe2c70c98f753306b70cbe9e0b5a"
   - github: digital-asset/hlint
-    commit: "951fdb6d28d7eed8ea1c7f3be69da29b61fcbe8f"
-    sha256: "f5fb4cf98cde3ecf1209857208369a63ba21b04313d570c41dffe9f9139a1d34"
-    size: 167936
+    commit: "2eb1c457c28f0dae87ca9987f1a45e2b0ced6aae"
```
Comment (cocreature) [inline stack-snapshot.yaml]: please bump this to the new commit after you merge your hlint PR.

File: compiler/damlc/daml-ide-core/dlint.yaml (hunk 3 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/daml-ide-core/dlint.yaml]: Ok

File: compiler/damlc/daml-ide-core/dlint.yaml (hunk 4 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/daml-ide-core/dlint.yaml]: Don't know. Looks like a mistake.

File: compiler/damlc/daml-ide-core/dlint.yaml (hunk 5 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/daml-ide-core/dlint.yaml]: Yes. Will remove all such notes.

File: compiler/damlc/daml-ide-core/dlint.yaml (hunk 6 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/daml-ide-core/dlint.yaml]: Ok

File: compiler/damlc/daml-ide-core/dlint.yaml (hunk 7 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/daml-ide-core/dlint.yaml]: Ok

File: compiler/damlc/daml-ide-core/dlint.yaml (hunk 8 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/daml-ide-core/dlint.yaml]: Ok

File: compiler/damlc/tests/daml-test-files/MaybeCompat.daml (hunk 9 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/tests/daml-test-files/MaybeCompat.daml]: Restored

File: stack-snapshot.yaml (hunk 10 above)
Comment (shayne-fletcher-da) [reply stack-snapshot.yaml]: Yes
